cursor_pagination.py

```diff
@@ -104,12 +80,39 @@ def page(self, first=None, last=None, after=None, before=None):
     def apply_cursor(self, cursor, queryset, reverse=False):
         position = self.decode_cursor(cursor)
 
-        is_reversed = self.ordering[0].startswith('-')
-        queryset = queryset.annotate(_cursor=Tuple(*[o.lstrip('-') for o in self.ordering]))
-        current_position = [Value(p, output_field=TextField()) for p in position]
-        if reverse != is_reversed:
-            return queryset.filter(_cursor__lt=Tuple(*current_position))
-        return queryset.filter(_cursor__gt=Tuple(*current_position))
+        filtering = Q()
+        q_acc = []
+
+        # this was previously implemented as tuple comparison done on postgres side
+        # Assume comparing 3-tuples a and b,
+        # the comparison a < b is equivalent to:
+        # (a.0 < b.0) || (a.0 == b.0 && (a.1 < b.1)) || (a.0 == b.0 && a.1 == b.1 && (a.2 < b.2))
+        # The expression above does not depend on short-circuit evalution support,
+        # which is usually unavailable on backend RDB
+
+        for position, ordering in zip(position, self.ordering):
+
+            is_reversed = ordering.startswith('-')
+            o = ordering.lstrip('-')
+
+            if reverse != is_reversed:
+                key_cmp = "{}__lt".format(o)
+            else:
+                key_cmp = "{}__gt".format(o)
+
+            key_eq = "{}__exact".format(o)
+
+            value = Value(position, output_field=TextField())
+
+            q = {key_cmp: value}
+            for q_eq in q_acc:
+                q.update(q_eq)
+
+            filtering |= Q(**q)
+
+            q_acc.append({key_eq: value})
+
+        return queryset.filter(filtering)
```

Comment:
This whole block feels like it has too much whitespace – take a look at the other methods for comparison.